pkg/tests/mappings/mappings_test.bzl

```diff
@@ -974,3 +975,53 @@ def mappings_unit_tests():
         "mappings_unit_tests",
         strip_prefix_test,
     )
+
+def _gen_manifest_test_main_impl(ctx):
+    ctx.actions.expand_template(
+        template = ctx.file._template,
+        output = ctx.outputs.out,
+        substitutions = {
+            "${EXPECTED}": ctx.attr.expected,
+            "${TARGET}": ctx.attr.target,
+            "${TEST_NAME}": ctx.attr.test_name,
+        },
+    )
+    return [
+        DefaultInfo(files = depset([ctx.outputs.out])),
+    ]
+
+_gen_manifest_test_main = rule(
+    implementation = _gen_manifest_test_main_impl,
+    attrs = {
+        "_template": attr.label(
+            default = Label("//tests/mappings:manifest_test_main.py.tpl"),
+            allow_single_file = True,
+        ),
+        "out": attr.output(mandatory = True),
+        "expected": attr.string(),
+        "target": attr.string(),
+        "test_name": attr.string(),
```

Comment:
Consider making these `mandatory` as well, given their use in `manifest_golden_test`.